### PR TITLE
mgr/dashboard: add option to resolve ip addr

### PR DESCRIPTION
**Pull request by s0nea**
Add the option `redirect_resolve_ip_addr` to the dashboard module.
If the option is set to `True`, try to resolve the IP address before redirecting from the passive to the active mgr instance.
If the option is set to `False`, follow the already known behavior.

Fixes: https://tracker.ceph.com/issues/56699
Signed-off-by: Tatjana Dehler <tdehler@suse.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/mgr/dashboard.rst [hunk 1]
```diff
@@ -1104,6 +1104,23 @@ code of standby dashboards. To do so you need to run the command::
 
   $ ceph config set mgr mgr/dashboard/standby_error_status_code 503
 
+Resolve IP address to hostname before redirect
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The redirect from the passive to the active dashboard is done via the IP
```
Comment (anthonyeleven) [inline doc/mgr/dashboard.rst:1110]: I think "from a standby to the active dashboard" might be better.

File: doc/mgr/dashboard.rst [hunk 2]
```diff
@@ -1104,6 +1104,23 @@ code of standby dashboards. To do so you need to run the command::
...
+address. This is done because resolving IP addresses to hostnames can be error
+prone in containerized environments. It is also the reason why the option is
+disabled by default.
+However, in some situation it might be helpful to redirect via the hostname.
```
Comment (anthonyeleven) [inline doc/mgr/dashboard.rst:1114]: situations

File: doc/mgr/dashboard.rst [hunk 3]
```diff
@@ -1104,6 +1104,23 @@ code of standby dashboards. To do so you need to run the command::
...
+For example if the configured TLS certificate matches the hostnames, only. To
```
Comment (anthonyeleven) [inline doc/mgr/dashboard.rst:1115]: only the hostnames

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

**Comment by s0nea** [comment]
@anthonyeleven updated. Thank you for your suggestions! :+1:

**Comment by s0nea** [comment]
jenkins test dashboard cephadm

**Review by anthonyeleven — APPROVED** [review]
Docs lgtm, code should be approved by someone else

File: doc/mgr/dashboard.rst [hunk 4]
```diff
@@ -1104,6 +1104,23 @@ code of standby dashboards. To do so you need to run the command::
...
+address. This is done because resolving IP addresses to hostnames can be error
+prone in containerized environments. It is also the reason why the option is
+disabled by default.
+However, in some situations it might be helpful to redirect via the hostname.
+For example if the configured TLS certificate matches only the hostnames. To
+activate the redirection via the hostname run the following command::
+
+  $ ceph config set mgr mgr/dashboard/redirect_resolve_ip_addr True
+
+You can disabled it again by::
```
Comment (epuertat) [inline doc/mgr/dashboard.rst:1120]:
```suggestion
You can disable it again by::
```

File: src/pybind/mgr/dashboard/module.py [hunk 5]
```diff
@@ -525,6 +529,21 @@ def default(self, *args, **kwargs):
                         return None
 
                     if active_uri:
+                        if module.get_module_option('redirect_resolve_ip_addr', False):
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/module.py:532]: No need to repeat the default, it's taken from the definition above:
```suggestion
                        if module.get_module_option('redirect_resolve_ip_addr'):
```

File: src/pybind/mgr/dashboard/module.py [hunk 6]
```diff
@@ -525,6 +529,21 @@ def default(self, *args, **kwargs):
...
+                            p_result = urlparse(active_uri)
+
+                            if p_result.hostname:
+                                hostname = socket.getfqdn(p_result.hostname)
+                                try:
+                                    ip = ipaddress.ip_address(hostname)
+                                except ValueError:
+                                    pass
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/module.py:540]: What's this dealing with? Maybe a short comment explaining why is needed.

**Review by epuertat — APPROVED** [review]
LGTM. Just a few suggestions.

**Comment by s0nea** [comment]
> LGTM. Just a few suggestions.

I updated the PR. @epuertat would you mind doing another review? - Thank you! :slightly_smiling_face:

**Review by epuertat — APPROVED** [review]
LGTM, just a small suggestion on a more concrete syntax.

**Comment by s0nea** [comment]
jenkins test make check

**Comment by s0nea** [comment]
jenkins test make check

**Comment by s0nea** [comment]
jenkins test dashboard

**Comment by s0nea** [comment]
jenkins test dashboard

**Comment by Pegonzal** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
@s0nea does this needs to be backported?

**Comment by s0nea** [comment]
> @s0nea does this needs to be backported?

Yes, I think so. I'll take a look. Thank you!

**Pull request merged**